In [30]:
from google.colab import drive        #Mounting the drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [31]:
import pandas as pd
filename = '/content/drive/MyDrive/CODING_DOJO/WEEK2/sales_predictions.csv'
sales_prediction_df=pd.read_csv(filename)         #importing and reading the data into panda dataframe
sales_prediction_df.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15         9.30  ...  Supermarket Type1          3735.1380
1           DRC01         5.92  ...  Supermarket Type2           443.4228
2           FDN15        17.50  ...  Supermarket Type1          2097.2700
3           FDX07        19.20  ...      Grocery Store           732.3800
4           NCD19         8.93  ...  Supermarket Type1           994.7052

[5 rows x 12 columns]

In [32]:
sales_prediction_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


***To clean the data, we will check for any duplicates and also the missing data in a data frame. We will delete the duplicated data. We will also check for unit consistencies. We will either eliminate or fill in the missing data.***

***Cleaning up data which can be used is essential to get good results.***

In [33]:
sales_prediction_df.shape

(8523, 12)

***Two columns have missing values.***

***Column Names: Item_Weight, Outlet_Size***


In [34]:
sales_prediction_df.isna().sum()        #getting the sum of the missing values in each column

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
                             ... 
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
Length: 12, dtype: int64

In [35]:
sales_prediction_df.duplicated().any()        #checking if there are any duplicated data

False

In [36]:
sales_prediction_df['Item_Weight'].value_counts()    #calculating most occured value in Item Weights column

12.150    86
17.600    82
13.650    77
11.800    76
15.100    68
          ..
7.560      2
9.420      1
5.400      1
6.520      1
7.685      1
Name: Item_Weight, Length: 415, dtype: int64

In [37]:
sales_prediction_df['Outlet_Size'].value_counts()       #calculating most occured value in Outlet Size column

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [38]:
sales_prediction_df['Item_Weight'].min()

4.555

In [39]:
sales_prediction_df['Item_Weight'].max()

21.35

In [40]:
Weight_mean=sales_prediction_df['Item_Weight'].mean()         #calculating the mean of Item Weight column
print(Weight_mean)

12.857645184136183


In [41]:
pd.set_option('display.max_rows', 10)
sales_prediction_df.groupby(['Item_Weight',  'Outlet_Size'])[['Outlet_Size']].count()         #checking if the grouping by Item Weight and Outlet Size 

Outlet_Size
Item_Weight Outlet_Size             
4.555       High                   1
            Medium                 1
            Small                  2
4.590       High                   1
            Medium                 1
...                              ...
21.250      Medium                 7
            Small                  5
21.350      High                   1
            Medium                 2
            Small                  2

[1082 rows x 1 columns]

In [42]:
sales_prediction_temp_df=sales_prediction_df.loc[(sales_prediction_df['Item_Weight']>12) & (sales_prediction_df['Item_Weight']<13)]  
sales_prediction_temp_df.groupby(['Item_Weight',  'Outlet_Size'])[['Outlet_Size']].count()  #checking the Outlet Size for the Item Weight between 12 and 13

Outlet_Size
Item_Weight Outlet_Size             
12.10       High                   6
            Medium                17
            Small                 12
12.15       High                  11
            Medium                26
...                              ...
12.80       Medium                 7
            Small                  6
12.85       High                   7
            Medium                16
            Small                 12

[27 rows x 1 columns]

In [43]:
Weight_missing=sales_prediction_df['Item_Weight'].isna()          #checking if the data missing is for the overlapping data points or different data points
Outlet_missing=sales_prediction_df['Outlet_Size'].isna()
sales_prediction_df.loc[(Weight_missing & Outlet_missing),('Item_Weight', 'Outlet_Size')]

Empty DataFrame
Columns: [Item_Weight, Outlet_Size]
Index: []

***Since there are no overlapping missing values, if we decide to eliminate the values, we will end up missing total of 3873 values which is about 45% of the data. So eliminating the data is not an option.***

In [44]:
sales_prediction_df['Item_Weight'].fillna(Weight_mean, inplace=True)   #Filling in the data in 'Item_Weight' Column by the way of filling in the mean value of the column

***'Item_Weight' column is filled in by the mean value of the column. Mean value is around 12.85. Mode / most occured value of the column is also 12.15. This value is close to the value of mean as well.***

In [45]:
sales_prediction_df.info()        #verifying that the missing values in column 'Item_Weight' are filled 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                8523 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [46]:
sales_prediction_df['Outlet_Size'].fillna('Medium', inplace=True)  
#Filling in the missing values in Outlet Size as the most occured value in the range of 12 to 13 Item Weight

***Based on the grouped analysis as per the Item Weight between 12 and 13. As the most occured values are around 12 and 13, Outlet Size estimation is based on the grouped analysis. As most occured Outlet Size value is medium, missing field is filled with that value.***

In [47]:
sales_prediction_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                8523 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                8523 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


***OBSERVATIONS***

***1. Data types for all the columns seems to be correct.***

***2.Missing Values only in 2 columns.***

***3.This data is from between the year 1985 and 2009.***

In [48]:
sales_prediction_df.mean()

Item_Weight                    12.857645
Item_Visibility                 0.066132
Item_MRP                      140.992782
Outlet_Establishment_Year    1997.831867
Item_Outlet_Sales            2181.288914
dtype: float64

In [54]:
sales_prediction_df['Item_Identifier'].value_counts()

FDG33    10
FDW13    10
NCB18     9
DRE49     9
NCF42     9
         ..
FDN52     1
FDY43     1
FDK57     1
FDE52     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64

In [56]:
sales_prediction_df['Outlet_Location_Type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

In [57]:
sales_prediction_df['Outlet_Size'].value_counts()

Medium    5203
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [58]:
sales_prediction_df['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

In [59]:
sales_prediction_df['Outlet_Establishment_Year'].value_counts()

1985    1463
1987     932
1999     930
1997     930
2004     930
2002     929
2009     928
2007     926
1998     555
Name: Outlet_Establishment_Year, dtype: int64